dhall-lsp-server/src/Dhall/LSP/Backend/Dhall.hs

```diff
@@ -1,34 +1,167 @@
+--   TODO: Split into FilePath -> FileIdentifier and URI -> Maybe FileIdentifier
+-- TODO: update comments
+fileIdentifierFromFilePath :: FilePath -> FileIdentifier
+fileIdentifierFromFilePath path =
+  let filename = Text.pack $ takeFileName path
+      directory = takeDirectory path
+      components = map Text.pack . reverse . splitDirectories $ directory
+  in FileIdentifier $ Dhall.Local Dhall.Absolute
+                        (Dhall.File (Dhall.Directory components) filename)
+
+fileIdentifierFromURI :: URI -> Maybe FileIdentifier
+fileIdentifierFromURI uri
+  | URI.uriScheme uri == "file:" = do
+    path <- LSP.Types.uriToFilePath . LSP.Types.Uri . Text.pack
+                  $ URI.uriToString id uri ""
+    return $ fileIdentifierFromFilePath path
+fileIdentifierFromURI uri
+  | otherwise = do
+    url <- Megaparsec.parseMaybe (Dhall.unParser Dhall.httpRaw) . Text.pack
+             $ URI.uriToString id uri ""
+    return $ FileIdentifier (Dhall.Remote url)
+
+-- | A well-typed expression.
+newtype WellTyped = WellTyped {fromWellTyped :: Expr Src X}
+
+-- | A fully normalised expression.
+newtype Normal = Normal {fromNormal :: Expr Src X}
+
+-- | A cache maps Dhall imports to fully normalised expressions. By reusing
+--   caches we can speeds up diagnostics etc. significantly!
+newtype Cache = Cache (Map.Map Dhall.Import (Dot.NodeId, Expr Src X))
+
+-- | The initial cache.
+emptyCache :: Cache
+emptyCache = Cache Map.empty
+
+-- | Cache a given normal expression.
+cacheExpr :: FileIdentifier -> Normal -> Cache -> Cache
+cacheExpr fileid (Normal expr) (Cache c) =
+  let unhashedImport = importFromFileIdentifier fileid
+      alpha = Dhall.alphaNormalize expr  -- we need to alpha-normalise before
+      hash = Dhall.hashExpression maxBound alpha  -- calculating the hash
+      hashedImport = hashedImportFromFileIdentifier fileid hash
+  in Cache $ Map.insert unhashedImport (Dot.userNodeId 0, expr)
+           $ Map.insert hashedImport (Dot.userNodeId 0, expr) c
 
-runDhall :: FilePath -> Text -> IO (Expr Src X)
-runDhall path = inputExprWithSettings dhallparams
+-- Construct the unhashed import corresponding to the given file.
+importFromFileIdentifier :: FileIdentifier -> Dhall.Import
+importFromFileIdentifier (FileIdentifier importType) =
+  Dhall.Import { importHashed = Dhall.ImportHashed Nothing importType,
+                 importMode = Dhall.Code }
+
+
+-- Construct the hashed import corresponding to the given file.
+hashedImportFromFileIdentifier :: FileIdentifier -> Digest SHA256 -> Dhall.Import
+hashedImportFromFileIdentifier (FileIdentifier importType) hash =
+  Dhall.Import { importHashed = Dhall.ImportHashed (Just hash) importType,
+                 importMode = Dhall.Code }
+
+-- | Invalidate any _unhashed_ imports of the given file. Hashed imports are
+--   kept around as per
+--   https://github.com/dhall-lang/dhall-lang/blob/master/standard/imports.md.
+--   Note to future self: this doesn't correctly invalidate reverse
```

Comment:
You should probably create an issue to track improving this invalidation behavior

Reply:
Will do!